core/vmx_ops.asm

```diff
@@ -0,0 +1,362 @@
+; POSSIBILITY OF SUCH DAMAGE.
+
+;
+; Detect architecture
+;
+%ifidn __OUTPUT_FORMAT__, elf32
+    %define __BITS__ 32
+    %define __CONV__ x32_cdecl
+%elifidn __OUTPUT_FORMAT__, win32
+    %define __BITS__ 32
+    %define __CONV__ x32_cdecl
+%elifidn __OUTPUT_FORMAT__, macho32
+    %define __BITS__ 32
+    %define __CONV__ x32_cdecl
+%elifidn __OUTPUT_FORMAT__, elf64
+    %define __BITS__ 64
+    %define __CONV__ x64_systemv
+%elifidn __OUTPUT_FORMAT__, win64
+    %define __BITS__ 64
+    %define __CONV__ x64_microsoft
+%elifidn __OUTPUT_FORMAT__, macho64
+    %define __BITS__ 64
+    %define __CONV__ x64_systemv
+%endif
+
+;
+; Describe calling convention
+;
+%ifidn __CONV__, x32_cdecl
+;
+; Although cdecl does not place arguments in registers, we simulate fastcall
+; by reading the first 2 stack arguments into the ecx/edx respectively.
+;
+    %define reg_arg1_16  cx
+    %define reg_arg1_32  ecx
+    %define reg_arg1     reg_arg1_32
+    %define reg_arg2_16  dx
+    %define reg_arg2_32  edx
+    %define reg_arg2     reg_arg2_32
+    %define reg_ret_16   ax
+    %define reg_ret_32   eax
+    %define reg_ret      reg_ret_32
+%elifidn __CONV__, x64_systemv
+    %define reg_arg1_16  di
+    %define reg_arg1_32  edi
+    %define reg_arg1_64  rdi
+    %define reg_arg1     reg_arg1_64
+    %define reg_arg2_16  si
+    %define reg_arg2_32  esi
+    %define reg_arg2_64  rsi
+    %define reg_arg2     reg_arg2_64
+    %define reg_ret_16   ax
+    %define reg_ret_32   eax
+    %define reg_ret_64   rax
+    %define reg_ret      reg_ret_64
+%elifidn __CONV__, x64_microsoft
+    %define reg_arg1_16  cx
+    %define reg_arg1_32  ecx
+    %define reg_arg1_64  rcx
+    %define reg_arg1     reg_arg1_64
+    %define reg_arg2_16  dx
+    %define reg_arg2_32  edx
+    %define reg_arg2_64  rdx
+    %define reg_arg2     reg_arg2_64
+    %define reg_ret_16   ax
+    %define reg_ret_32   eax
+    %define reg_ret_64   rax
+    %define reg_ret      reg_ret_64
+%endif
+
+;
+; Helpers
+;
+
+; Macro: function
+; Declares a function. Arguments:
+; - %1  Name of the function
+; - %2  Number of arguments
+;
+%macro function 2
+    global %1
+    %1:
+%ifidn __CONV__, x32_cdecl
+    %if %2 >= 3
+        %error "Unsupported number of arguments"
+    %else
+        %if %2 >= 1
+            mov reg_arg1, [esp + 0x4]
+        %endif
+        %if %2 >= 2
+            mov reg_arg2, [esp + 0x8]
+        %endif
+    %endif
+%endif
+%endmacro
+
+section .text
+
+struc qword_struct
+    .lo      resd 1
+    .hi      resd 1
+endstruc
+
+struc vcpu_state
+    ._rax    resq 1
+    ._rcx    resq 1
+    ._rdx    resq 1
+    ._rbx    resq 1
+    ._rsp    resq 1
+    ._rbp    resq 1
+    ._rsi    resq 1
+    ._rdi    resq 1
+    ._r8     resq 1
+    ._r9     resq 1
+    ._r10    resq 1
+    ._r11    resq 1
+    ._r12    resq 1
+    ._r13    resq 1
+    ._r14    resq 1
+    ._r15    resq 1
+endstruc
+
+%macro vmx_check 0
+    pushf
```

Comment:
`pushf => pushfw`

This turns out to be the root cause of the host freeze issue! Although the `pushf` mnemonic is used by the Intel SDM for the 16-bit (operand size) variant of the instruction, it is still ambiguous, and in this case gets assembled into `pushfq` by NASM for win64. `pushfw`, on the other hand, is unambiguous.